modulemd/modulemd-compression.c

```diff
@@ -0,0 +1,277 @@
+           g_str_has_suffix (filename, ".txt"))
+    {
+      return MODULEMD_COMPRESSION_TYPE_NO_COMPRESSION;
+    }
+
+#ifdef HAVE_LIBMAGIC
+  /* No known suffix? Try using libmagic from file-utils */
+  const char *mime_type;
+  g_auto (magic_t) magic = NULL;
+  int magic_fd = dup (fd);
+
+  magic = magic_open (MAGIC_MIME);
+  if (magic == NULL)
+    {
+      g_set_error (error,
+                   MODULEMD_ERROR,
+                   MODULEMD_ERROR_MAGIC,
+                   "magic_open() failed: Cannot allocate the magic cookie");
+      return MODULEMD_COMPRESSION_TYPE_DETECTION_FAILED;
+    }
+
+  if (magic_load (magic, NULL) == -1)
+    {
+      g_set_error (error,
+                   MODULEMD_ERROR,
+                   MODULEMD_ERROR_MAGIC,
+                   "magic_load() failed: %s",
+                   magic_error (magic));
+      return MODULEMD_COMPRESSION_TYPE_DETECTION_FAILED;
+    }
+
+  mime_type = magic_descriptor (magic, magic_fd);
+  close(magic_fd);
+
+  if (mime_type)
+    {
+      g_debug (
+        "%s: Detected mime type: %s (%s)", __func__, mime_type, filename);
+
+      if (g_str_has_prefix (mime_type, "application/x-gzip") ||
+          g_str_has_prefix (mime_type, "application/gzip") ||
+          g_str_has_prefix (mime_type, "application/gzip-compressed") ||
+          g_str_has_prefix (mime_type, "application/gzipped") ||
+          g_str_has_prefix (mime_type, "application/x-gzip-compressed") ||
+          g_str_has_prefix (mime_type, "application/x-compress") ||
+          g_str_has_prefix (mime_type, "application/x-gzip") ||
+          g_str_has_prefix (mime_type, "application/x-gunzip") ||
+          g_str_has_prefix (mime_type, "multipart/x-gzip"))
+        {
+          type = MODULEMD_COMPRESSION_TYPE_GZ_COMPRESSION;
+        }
+
+      else if (g_str_has_prefix (mime_type, "application/x-bzip2") ||
+               g_str_has_prefix (mime_type, "application/x-bz2") ||
+               g_str_has_prefix (mime_type, "application/bzip2") ||
+               g_str_has_prefix (mime_type, "application/bz2"))
+        {
+          type = MODULEMD_COMPRESSION_TYPE_BZ2_COMPRESSION;
+        }
+
+      else if (g_str_has_prefix (mime_type, "application/x-xz"))
+        {
+          type = MODULEMD_COMPRESSION_TYPE_XZ_COMPRESSION;
+        }
+
+      else if (g_str_has_prefix (mime_type, "text/plain") ||
+               g_str_has_prefix (mime_type, "text/x-yaml") ||
+               g_str_has_prefix (mime_type, "application/x-yaml"))
+        {
+          type = MODULEMD_COMPRESSION_TYPE_NO_COMPRESSION;
+        }
+    }
+  else
+    {
+      g_set_error (error,
+                   MODULEMD_ERROR,
+                   MODULEMD_ERROR_MAGIC,
+                   "mime_type() detection failed: %s",
+                   magic_error (magic));
+      return MODULEMD_COMPRESSION_TYPE_DETECTION_FAILED;
+    }
+  /* Reset the file descriptor to the start of the file, if it has moved */
+  lseek (fd, 0, SEEK_SET);
```

Comment:
I'd suggest moving this to right after `close(magic_fd);` above--in case the caller decides to proceed with reading from `fd` (perhaps treating it as uncompressed) even when `MODULEMD_COMPRESSION_TYPE_DETECTION_FAILED` is returned.

Reply:
Good idea, done.